src/Architectures.jl

```diff
@@ -54,46 +66,61 @@ On single-process, non-distributed systems, return `arch`.
 child_architecture(arch) = arch
 
 array_type(::CPU) = Array
-array_type(::GPU) = CuArray
+array_type(::CUDAGPU) = CuArray
+array_type(::ROCmGPU) = ROCArray
 
 arch_array(::CPU, a::Array)   = a
 arch_array(::CPU, a::CuArray) = Array(a)
-arch_array(::GPU, a::Array)   = CuArray(a)
-arch_array(::GPU, a::CuArray) = a
-
+arch_array(::CPU, a::ROCArray) = Array(a)
+arch_array(::CUDAGPU, a::Array) = CuArray(a)
+arch_array(::ROCmGPU, a::Array) = ROCArray(a)
+arch_array(::CUDAGPU, a::CuArray) = a
+arch_array(::ROCmGPU, a::ROCArray) = a
 arch_array(::CPU, a::BitArray) = a
-arch_array(::GPU, a::BitArray) = CuArray(a)
+arch_array(::CUDAGPU, a::BitArray) = CuArray(a)
+arch_array(::ROCmGPU, a::BitArray) = ROCArray(a)
+
+arch_array(::CUDAGPU, a::SubArray{<:Any, <:Any, <:CuArray}) = a
+arch_array(::ROCmGPU, a::SubArray{<:Any, <:Any, <:ROCArray}) = a
 
-arch_array(::GPU, a::SubArray{<:Any, <:Any, <:CuArray}) = a
 arch_array(::CPU, a::SubArray{<:Any, <:Any, <:CuArray}) = Array(a)
+arch_array(::CUDAGPU, a::SubArray{<:Any, <:Any, <:Array}) = CuArray(a)
+arch_array(::ROCmGPU, a::SubArray{<:Any, <:Any, <:Array}) = ROCArray(a)
 
-arch_array(::GPU, a::SubArray{<:Any, <:Any, <:Array}) = CuArray(a)
 arch_array(::CPU, a::SubArray{<:Any, <:Any, <:Array}) = a
 
 arch_array(::CPU, a::AbstractRange) = a
 arch_array(::CPU, ::Nothing)   = nothing
 arch_array(::CPU, a::Number)   = a
 arch_array(::CPU, a::Function) = a
 
-arch_array(::GPU, a::AbstractRange) = a
-arch_array(::GPU, ::Nothing)   = nothing
-arch_array(::GPU, a::Number)   = a
-arch_array(::GPU, a::Function) = a
+arch_array(::CUDAGPU, a::AbstractRange) = a
+arch_array(::CUDAGPU, ::Nothing)   = nothing
+arch_array(::CUDAGPU, a::Number)   = a
+arch_array(::CUDAGPU, a::Function) = a
+
+arch_array(::ROCmGPU, a::AbstractRange) = a
+arch_array(::ROCmGPU, ::Nothing)   = nothing
+arch_array(::ROCmGPU, a::Number)   = a
+arch_array(::ROCmGPU, a::Function) = a
 
 arch_array(arch::CPU, a::OffsetArray) = OffsetArray(arch_array(arch, a.parent), a.offsets...)
-arch_array(arch::GPU, a::OffsetArray) = OffsetArray(arch_array(arch, a.parent), a.offsets...)
+arch_array(arch::CUDAGPU, a::OffsetArray) = OffsetArray(arch_array(arch, a.parent), a.offsets...)
+arch_array(arch::ROCmGPU, a::OffsetArray) = OffsetArray(arch_array(arch, a.parent), a.offsets...)
```

Comment:
these two can be combined; they are the same!

```suggestion
arch_array(arch::GPU{D}, a::OffsetArray) where D = OffsetArray(arch_array(arch, a.parent), a.offsets...)
```